test/extensions/quic_listeners/quiche/active_quic_listener_test.cc

```diff
@@ -84,102 +184,81 @@ class ActiveQuicListenerTest : public testing::TestWithParam<Network::Address::I
-                                                 nullptr, *listen_socket_->localAddress());
-  ASSERT_EQ(encrypted_packet->length(), send_rc.rc_);
-
-  EXPECT_CALL(listener_config_, filterChainManager());
-  EXPECT_CALL(filter_chain_manager_, findFilterChain(_));
-  EXPECT_CALL(filter_chain_, networkFilterFactories());
-  EXPECT_CALL(listener_config_, filterChainFactory());
-  EXPECT_CALL(listener_config_.filter_chain_factory_, createNetworkFilterChain(_, _));
-  EXPECT_CALL(*read_filter_, onNewConnection())
-      // Stop iteration to avoid calling getRead/WriteBuffer().
-      .WillOnce(Invoke([this]() {
-        dispatcher_->exit();
-        return Network::FilterStatus::StopIteration;
-      }));
-  EXPECT_CALL(network_connection_callbacks_, onEvent(Network::ConnectionEvent::Connected));
-
-  dispatcher_->run(Event::Dispatcher::RunType::Block);
-
-  Buffer::InstancePtr result_buffer(new Buffer::OwnedImpl());
-  const uint64_t bytes_to_read = 11;
-  uint64_t bytes_read = 0;
-  int retry = 0;
-
-  do {
-    Api::IoCallUint64Result result =
-        result_buffer->read(client_socket_->ioHandle(), bytes_to_read - bytes_read);
-
-    if (result.ok()) {
-      bytes_read += result.rc_;
-    } else if (retry == 10 || result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
-      break;
-    }
+  ConfigureMocks(/* connection_count = */ 1);
+  GenerateCHLO(quic::test::TestConnectionId(1));
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  ReadFromClientSockets();
+}
 
-    if (bytes_read == bytes_to_read) {
-      break;
-    }
+TEST_P(ActiveQuicListenerTest, ProcessBufferedChlos) {
+  EnvoyQuicDispatcher* const envoy_quic_dispatcher =
+      ActiveQuicListenerPeer::quic_dispatcher(*quic_listener_);
+  quic::QuicBufferedPacketStore* const buffered_packets =
+      quic::test::QuicDispatcherPeer::GetBufferedPackets(envoy_quic_dispatcher);
+
+  ConfigureMocks(ActiveQuicListener::kNumSessionsToCreatePerEpoll + 2);
+
+  // Generate one more CHLO than can be processed immediately.
+  for (size_t i = 1; i <= ActiveQuicListener::kNumSessionsToCreatePerEpoll + 1; ++i) {
+    GenerateCHLO(quic::test::TestConnectionId(i));
+  }
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+
+  // The first kNumSessionsToCreatePerEpoll CHLOs are processed,
+  // the last one is buffered.
+  for (size_t i = 1; i <= ActiveQuicListener::kNumSessionsToCreatePerEpoll; ++i) {
+    EXPECT_FALSE(buffered_packets->HasBufferedPackets(quic::test::TestConnectionId(i)));
+  }
+  EXPECT_TRUE(buffered_packets->HasBufferedPackets(
+      quic::test::TestConnectionId(ActiveQuicListener::kNumSessionsToCreatePerEpoll + 1)));
+  EXPECT_TRUE(buffered_packets->HasChlosBuffered());
 
-    retry++;
-    ::usleep(10000);
-  } while (true);
-  // TearDown() will close the connection.
-  EXPECT_CALL(network_connection_callbacks_, onEvent(Network::ConnectionEvent::LocalClose));
+  // Generate more data to trigger a socket read during the next event loop.
```

Comment:
This reminds me that we are missing one feature: if any CHLO gets buffered, we should artifacts a READ event instead of relying on incoming traffic to trigger onReadReady(). This needs to be added at the end of UdpListenerImpl::onSocketEvent() by observing the return value of onData() propagated to  Utility::readPacketsFromSocket(). And it may share the same logic in UdpListenerImpl to artifact a READ event when we limit number of reads per event loop in the future. Please add a TODO in onData() to return an indicator of such need. Or modify onData() and processPacket() in this PR to return a bool but without handling the return value in readPacketsFromSocket().

Reply:
That sounds justified.

How are you planning to trigger ProcessBufferedCHLOs() in the next event loop if there is no incoming traffic?  What you are describing sounds more like doing something at the end of onSocketEvent() in the same event loop.  Maybe we'll need to find a place for it outside socket events altogether.

In any case, I'm adding the TODO here for the time being.

Reply:
SG